akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/ReplicatorMessageResilienceSpec.scala

```diff
@@ -0,0 +1,169 @@
+  private val replicator = system.actorOf(Replicator.props(ReplicatorSettings(system).withGossipInterval(3.seconds)))
+
+  private val timeout = 5.seconds.dilated
+
+  private val Key = ORSetKey[String]("load")
+
+  private val maximumFrameSize = RARP(system).provider.remoteSettings.Artery.Advanced.MaximumFrameSize
+
+  private val N = (maximumFrameSize / 1024) / 2 // 32768/1024=32 / 2 runs
+
+  def join(from: RoleName, to: RoleName): Unit = {
+    runOn(from) {
+      cluster.join(node(to).address)
+    }
+
+    enterBarrier(from.name + "-joined")
+  }
+
+  private val random = new scala.util.Random()
+
+  def payload(size: Int) = random.nextString(size)
+
+  def update(): Unit = {
+
+    val each = (maximumFrameSize / 1024) - 1
+
+    (0 until N).foreach { _ =>
+      replicator ! Update(Key, ORSet(), WriteLocal)(_ :+ payload(each))
+    }
+
+    receiveN(N).map(_.getClass).toSet shouldEqual Set(classOf[UpdateSuccess[_]])
+  }
+
+  "Replicator in cluster with network issues" must {
+
+    "setup: replicate initial data" in {
+      join(first, first)
+      join(second, first)
+      join(third, first)
+      join(fourth, first)
+
+      within(10.seconds) {
+        awaitAssert {
+          replicator ! GetReplicaCount
+          expectMsg(ReplicaCount(roles.size))
+        }
+      }
+      enterBarrier("all-joined")
+
+      runOn(first, second, third) {
+        update()
+      }
+      enterBarrier("update-1")
+
+      within(timeout) {
+        awaitAssert {
+          replicator ! Get(Key, ReadLocal)
+          expectMsgType[GetSuccess[ORSet[_]]].dataValue.size shouldEqual (roles.size - 1) * N
+        }
+      }
+      enterBarrier("replicated-1")
+    }
+
+    "replicate changes without OversizedPayloadException" in {
+
+      runOn(first, second, third) {
+        update()
+      }
+      enterBarrier("update-2")
+
+      within(5.seconds) {
+        awaitAssert {
+          replicator ! Get(Key, ReadLocal)
+          expectMsgType[GetSuccess[ORSet[_]]].dataValue.size shouldEqual (roles.size - 1) * (N * 2)
+        }
+      }
+      enterBarrier("replicated-2")
+
+    }
+
+    "raise OversizedPayloadException for Replicator.Internal.Gossip" in {
+      // TODO testConductor.blackhole this for the realistic model of network connectivity.
+      // This just points it out so far as a placeholder.
+      // Gossip and delta propagation builds up. When these messages are triggered for replication,
+      // the byte size has exceeded the max, thus it currently raises the error and the data is dropped.
+
+      val oversized = maximumFrameSize + 1
+
+      runOn(first) {
+        within(timeout * 5) {
+          awaitAssert {
+
+            EventFilter
+              .error(
+                start = "Failed to serialize oversized message [akka.cluster.ddata.Replicator$Internal$Gossip]",
+                occurrences = 1)
+              .intercept {
+
+                replicator ! Update(Key, ORSet(), WriteLocal)(_ :+ payload(oversized))
```

Comment:
This will never work (if deltas are disabled) since the ORSet entry can't be slit up into smaller pieces.

The test should probably use several different keys and add entries making the total size of all of the top level entries larger than the maximumFrameSize, but each ORSet itself is smaller than maximumFrameSize.

Then the test can try the full gossip by joining fourth later. Then fourth will sync with some of the others via Gossip.

If I understand the ticket correctly it is about automatically and dynamically finding a `max-delta-elements` based on the payload size, instead of configuring a fixed `max-delta-elements`.